src/librustc_privacy/lib.rs

```diff
@@ -425,14 +427,15 @@ impl<'a, 'tcx, 'v> Visitor<'v> for PrivacyVisitor<'a, 'tcx> {
                 let method = self.tcx.tables.borrow().method_map[&method_call];
                 self.check_method(expr.span, method.def_id);
             }
-            hir::ExprStruct(..) => {
+            hir::ExprStruct(_, ref fields, _) => {
                 let adt = self.tcx.expr_ty(expr).ty_adt_def().unwrap();
                 let variant = adt.variant_of_def(self.tcx.expect_def(expr.id));
                 // RFC 736: ensure all unmentioned fields are visible.
                 // Rather than computing the set of unmentioned fields
                 // (i.e. `all_fields - fields`), just check them all.
-                for field in &variant.fields {
-                    self.check_field(expr.span, adt, field);
+                for (field, field_span) in variant.fields.iter()
+                                           .zip(fields.iter().map(|f| f.span)) {
+                    self.check_field(field_span, adt, field);
```

Comment:
@mikhail-m1: The problem, I believe, is that `fields` contains less fields than `variant.fields`, which means that not all fields are iterated over.

Also I don't think it's guaranteed that the fields in `fields` and `variant.fields` occur in the same order - my intuition would be that `variant.fields` is in the order the fields were defined and `fields` is in the order the fields were written in the struct literal.